devito/passes/clusters/buffering.py

```diff
@@ -1,25 +1,26 @@
-from collections import OrderedDict, defaultdict, namedtuple
-from itertools import combinations
+from collections import defaultdict, namedtuple
 from functools import cached_property
+from itertools import chain
 
+from sympy import S
 import numpy as np
 
-from devito.ir import (Cluster, Forward, GuardBound, Interval, IntervalGroup,
-                       IterationSpace, AFFINE, PARALLEL, Queue, Vector,
-                       lower_exprs, vmax, vmin)
+from devito.ir import (Cluster, Backward, Forward, GuardBound, Interval,
+                       IntervalGroup, IterationSpace, Properties, Queue, Vector,
+                       InitArray, lower_exprs, vmax, vmin)
 from devito.exceptions import InvalidOperator
 from devito.logger import warning
 from devito.passes.clusters.utils import is_memcpy
-from devito.symbolics import IntDiv, retrieve_function_carriers, uxreplace
-from devito.tools import (Bunch, DefaultOrderedDict, Stamp, as_tuple,
-                          filter_ordered, flatten, is_integer, timed_pass)
-from devito.types import Array, CustomDimension, Dimension, Eq, ModuloDimension
+from devito.symbolics import IntDiv, retrieve_functions, uxreplace
+from devito.tools import (Stamp, as_mapper, as_tuple, filter_ordered, frozendict,
+                          flatten, is_integer, timed_pass)
+from devito.types import Array, CustomDimension, Eq, ModuloDimension
 
 __all__ = ['buffering']
 
 
 @timed_pass()
-def buffering(clusters, callback, sregistry, options, **kwargs):
+def buffering(clusters, key, sregistry, options, **kwargs):
```

Comment:
Note for reviewers: unlike `passes/clusters/asynchrony.py`, which for a good 85-90% is the same code but refactored, this pass has been totally revamped and the result is, imho, vastly better (more robust, easier to read, dramatically easier to understand and mofify) code